### PR TITLE
FOGL-4569 Add API to fetch asset struture for all buffered assets

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Signed-off-by: Mark Riddoch <mark@dianomic.com>

### CODE REVIEW

File: python/fledge/services/core/api/browser.py [hunk 1]
```diff
@@ -608,3 +612,52 @@ async def asset_readings_with_bucket_size(request: web.Request) -> web.Response:
...
+                    if value == "True" or value == "False":
+                        datapoint[name] = "boolean"
+                    else:
+                        metadata[name] = value
+                elif type(value) == int:
+                    datapoint[name] = "integer"
+                elif type(value) == float:
+                    datapoint[name] = "float"
+            if len(metadata) > 0:
+                asset_json[code] = { 'datapoint' : datapoint, 'metadata' : metadata }
```
Comment (ashish-jabble) [inline python/fledge/services/core/api/browser.py:654]: Fix whitespace after '{' and before ':', '}'

File: python/fledge/services/core/api/browser.py [hunk 2]
```diff
@@ -608,3 +612,52 @@ async def asset_readings_with_bucket_size(request: web.Request) -> web.Response:
...
+            for name, value in asset['reading'].items():
+                if type(value) == str:
+                    if value == "True" or value == "False":
+                        datapoint[name] = "boolean"
+                    else:
+                        metadata[name] = value
+                elif type(value) == int:
+                    datapoint[name] = "integer"
+                elif type(value) == float:
+                    datapoint[name] = "float"
```
Comment (ashish-jabble) [inline python/fledge/services/core/api/browser.py:652]: if type(value) other than `str, int, float`?

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
Remove error logs from code or with level debug. Fix unit tests.

File: python/fledge/services/core/api/browser.py [hunk 3]
```diff
@@ -64,6 +67,7 @@ def setup(app):
     app.router.add_route('GET', '/fledge/asset/{asset_code}/{reading}/series', asset_averages)
     app.router.add_route('GET', '/fledge/asset/{asset_code}/bucket/{bucket_size}', asset_datapoints_with_bucket_size)
     app.router.add_route('GET', '/fledge/asset/{asset_code}/{reading}/bucket/{bucket_size}', asset_readings_with_bucket_size)
+    app.router.add_route('GET', '/fledge/structure/asset', asset_structure)
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:70]: That will not work as "structure" could be an asset code

File: python/fledge/services/core/api/browser.py (hunk 2 above)
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:652]: Those are the only things that are needed.

File: python/fledge/services/core/api/browser.py [hunk 4]
```diff
@@ -608,3 +612,52 @@ async def asset_readings_with_bucket_size(request: web.Request) -> web.Response:
...
+    payload = PayloadBuilder().GROUP_BY("asset_code").payload()
+
+    results = {}
+    try:
+        _readings = connect.get_readings_async()
+        results = await _readings.query(payload)
+        _logger.error("Strcuture query run")
+        assets = results['rows']
+        asset_json = {}
+        for asset in assets:
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:635]: Where? I don't see another definition?

File: python/fledge/services/core/api/browser.py [hunk 5]
```diff
@@ -608,3 +612,52 @@ async def asset_readings_with_bucket_size(request: web.Request) -> web.Response:
...
+                    else:
+                        metadata[name] = value
+                elif type(value) == int:
+                    datapoint[name] = "integer"
+                elif type(value) == float:
+                    datapoint[name] = "float"
+            if len(metadata) > 0:
+                asset_json[code] = { 'datapoint' : datapoint, 'metadata' : metadata }
+            else:
+                asset_json[code] = { 'datapoint' : datapoint }
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:652]: I prefer better spacing to make it readable

File: python/fledge/services/core/api/browser.py (hunk 1 above)
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:650]: As above

File: python/fledge/services/core/api/browser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashish-jabble) [reply python/fledge/services/core/api/browser.py:652]: These are PEP 8 coding style violation

File: python/fledge/services/core/api/browser.py (hunk 1 above)
Comment (ashish-jabble) [reply python/fledge/services/core/api/browser.py:650]: These are PEP 8 coding style violation

File: python/fledge/services/core/api/browser.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashish-jabble) [reply python/fledge/services/core/api/browser.py:635]: See we have https://github.com/fledge-iot/fledge/blob/49bf1795e9b2f73cd8523fa9de32888bda720698/python/fledge/services/core/api/browser.py#L132

File: python/fledge/services/core/api/browser.py [hunk 6]
```diff
@@ -654,9 +650,11 @@ async def asset_structure(request):
                 asset_json[code] = { 'datapoint' : datapoint, 'metadata' : metadata }
             else:
                 asset_json[code] = { 'datapoint' : datapoint }
-            _logger.error("%s : %s", code, asset_json[code])
     except KeyError:
-        raise web.HTTPBadRequest(reason=results['message'])
+        msg = results['message']
+        raise web.HTTPBadRequest(reason=results['message'], body=json.dumps({"message": msg}))
+    except Exception as e:
+        raise web.HTTPInternalServerError(reason=str(e))
```
Comment (ashish-jabble) [inline python/fledge/services/core/api/browser.py:657]: Add JSON body as well

File: python/fledge/services/core/api/browser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:652]: This is a really bad coding style in my opinion

File: python/fledge/services/core/api/browser.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply python/fledge/services/core/api/browser.py:635]: Thanks, missed that one

**Review by ashish-jabble — APPROVED** [review]
LGTM

**Pull request merged**